flang/lib/Lower/ConvertType.cpp

```diff
@@ -221,12 +223,22 @@ struct TypeBuilder {
         std::vector<std::pair<std::string, mlir::Type>> cs;
         auto &symbol = tySpec->typeSymbol();
         auto rec = fir::RecordType::get(context, toStringRef(symbol.name()));
-        // TODO: use Fortran::semantics::ComponentIterator to go through
-        // components. or use similar mechanism. We probably want to go through
-        // the Ordered components.
-        TODO("lower derived type to fir types");
+        Fortran::semantics::OrderedComponentIterator components(*tySpec);
+        for (auto iter = components.begin(), iend = components.end();
+             iter != iend; ++iter) {
+          auto &field = *iter;
+          cs.emplace_back(field.name().ToString(), genSymbolType(field));
+        }
+        for (auto iter : tySpec->parameters())
+          if (iter.second.isLen()) {
+            ps.emplace_back(
+                iter.first.ToString(),
+                /* FIXME - is the type available? Fake with index type. */
```

Comment:
`DerivedTypeSpec` represents an instance of of a derived type. It does not store the types of the type parameters, but their values (assumed, deferred, explicit expr). The type information is stored in the Symbol that represents the type definition. You can get the type definition from a `DerivedTypeSpec` with the `typeSymbol()` member function.
All that to say, I think you can use something like:
```
for (const auto& param: Fortran::semantics::OrderParameterDeclarations(tySpec->typeSymbol())
   if (param.get<TypeParamDetails>().attr() == Fortran::common::TypeParamAttr::Len))
       ps.emplace_back(param.name.ToString(), genSymbolType(param));
      
```

See `OrderParameterDeclarations` declaration [here ](https://github.com/llvm/llvm-project/blob/ebe74d9592d097501f376c2086e58d35aa318896/flang/include/flang/Semantics/tools.h#L207).

Reply:
Thanks!